### PR TITLE
update fetcher to retrieve latest version

### PR DESCRIPTION
**Pull request by pkwarren**
Maven versions can be returned out of order (the default sort is by score). Update the query to determine the latest semver version from the API (ignoring pre-release versions).

Additionally, fetch the latest patch version for each major+minor (we can update the utility in the future to use this to update to later patch versions of plugins we already ship).

### CODE REVIEW

File: internal/fetchclient/fetchclient.go [hunk 1]
```diff
@@ -172,22 +181,24 @@ func (c *Client) fetchMaven(ctx context.Context, group string, name string) (str
...
-		return "", err
+		return nil, err
 	}
 	defer response.Body.Close()
 
 	var data struct {
 		Response struct {
-			Docs []struct {
+			NumFound int `json:"numFound"`
+			Start    int `json:"start"`
```
Comment (pkwarren) [inline internal/fetchclient/fetchclient.go:200]: Added these for future use (if we want to paginate over all versions). Right now scanning the first 20 results works ok.

File: internal/fetchclient/fetchclient.go [hunk 2]
```diff
@@ -196,12 +207,36 @@ func (c *Client) fetchMaven(ctx context.Context, group string, name string) (str
...
+		version := doc.Version
+		if !strings.HasPrefix(version, "v") {
+			version = "v" + version
+		}
+		if !semver.IsValid(version) || semver.Prerelease(version) != "" {
+			continue
+		}
+		version = semver.Canonical(version)
+		if results.latestVersion == "" || semver.Compare(results.latestVersion, version) < 0 {
+			results.latestVersion = version
```
Comment (pkwarren) [inline internal/fetchclient/fetchclient.go:229]: Keep track of the latest available version.

File: internal/fetchclient/fetchclient.go [hunk 3]
```diff
@@ -196,12 +207,36 @@ func (c *Client) fetchMaven(ctx context.Context, group string, name string) (str
...
+		}
+		release := semver.MajorMinor(version)
+		if results.latestPatchVersions[release] == "" || semver.Compare(results.latestPatchVersions[release], version) < 0 {
+			results.latestPatchVersions[release] = version
```
Comment (pkwarren) [inline internal/fetchclient/fetchclient.go:233]: Keep track of the latest version for a given major + minor release. We could use this in a follow up PR to update to the latest patch version of a plugin we're already shipping.

**Comment by pkwarren** [comment]
This works around a bug in #131 where we aren't correctly detecting the latest version of the grpc-java plugin.

**Review by smallsamantha — APPROVED** [review]

**Comment by mfridman** [comment]
> Maven versions can be returned out of order (the default sort is by score). Update the query to determine the latest semver version from the API (ignoring pre-release versions).

The changes look good to me, thanks for fixing this up. I'm curious, where does it say the default sort is by score? Afaics the docs say:

> Returns sorted list of all versions of an artifact.

Ref, the [second example](https://central.sonatype.org/search/rest-api-guide/).

**Pull request merged**